### PR TITLE
explore db with sqlalchemy and generate schema of the db

### LINKED ISSUE #33: Generate db schema with pagai

### PR DESCRIPTION
**Pull request by elsiehoffet-94**
Closes #33

### CODE REVIEW

File: pagai/services/database_explorer.py [hunk 1]
```diff
@@ -102,3 +103,21 @@ def explore(self, table: str, limit, schema=None, filters=[]):
                 raise OperationOutcome(e)
         except Exception as e:
             raise OperationOutcome(e)
+
+    def get_db_schema(self, owner=None):
+        db_schema = defaultdict(lambda: defaultdict(lambda: list()))
```
Comment (Jasopaum) [inline pagai/services/database_explorer.py:108]: nit: in the last layer, defaultdict(list) is sufficient

File: pagai/services/database_explorer.py [hunk 2]
```diff
@@ -104,20 +104,34 @@ def explore(self, table: str, limit, schema=None, filters=[]):
...
-                result = connection.execute(
-                    "SELECT table_schema, table_name, column_name FROM information_schema.columns;"
-                )
-                for row in result:
-                    db_schema[row["table_schema"]][row["table_name"]].append(row["column_name"])
-        return db_schema[owner] if owner else db_schema
+            sql_query = text(f"select schema_name as owners from information_schema.schemata;")
+        with self._sql_engine.connect() as connection:
+            result = connection.execute(sql_query).fetchall()
+        return {"owners": [r["owners"] for r in result]}
```
Comment (simonvadee) [inline pagai/services/database_explorer.py:117]: I think you can use a list instead of a dict here

File: pagai/views.py [hunk 3]
```diff
@@ -144,20 +144,43 @@ def explore(resource_id, table):
         raise OperationOutcome(e)
 
 
+@api.route("/get_owners/", methods=["POST"])
```
Comment (simonvadee) [inline pagai/views.py:147]: this is not optimal but I can't think of an other secure solution

File: pagai/views.py [hunk 4]
```diff
@@ -144,6 +144,54 @@ def explore(resource_id, table):
...
+    except OperationalError as e:
+        if "could not connect to server" in str(e):
+            raise OperationOutcome(f"Could not connect to the database: {e}")
+        else:
+            raise OperationOutcome(e)
+    except Exception as e:
+        raise OperationOutcome(e)
+
+
+@api.route("/get_db_schema/", methods=["POST"])
```
Comment (simonvadee) [inline pagai/views.py:169]: you need to remove the trailing /

File: pagai/views.py (hunk 3 above)
Comment (simonvadee) [reply pagai/views.py:147]: you need to remove the trailing "/"

**Pull request merged**